File: core/src/main/java/com/gitlab/aecsocket/sokol/core/impl/AbstractComponent.java
package com.gitlab.aecsocket.sokol.core.impl;

import com.gitlab.aecsocket.sokol.core.*;
import com.gitlab.aecsocket.sokol.core.registry.Keyed;
import com.gitlab.aecsocket.sokol.core.stat.StatIntermediate;

import java.util.*;

public abstract class AbstractComponent<
        C extends AbstractComponent<C, S, F, N>,
        S extends Slot,
        F extends Feature<? extends FeatureInstance<N>, N>,
        N extends Node.Scoped<N, ?, ?>
> implements Component.Scoped<C, S, F, N> {
    protected final String id;
    protected final Set<String> tags;
    protected final Map<String, S> slots;
    protected final Map<String, F> features;
    protected final StatIntermediate stats;

    public AbstractComponent(String id, Set<String> tags, Map<String, S> slots, Map<String, F> features, StatIntermediate stats) {
        Keyed.validate(id);
        this.id = id;
        this.tags = tags;
        this.slots = slots;
        this.features = features;
        this.stats = stats;
    }

    @Override public String id() { return id; }

    @Override public Set<String> tags() { return new HashSet<>(tags); }
    @Override public boolean tagged(String tag) { return tags.contains(tag); }

    @Override public Map<String, S> slots() { return new HashMap<>(slots); }
    @Override public Optional<S> slot(String key) { return Optional.ofNullable(slots.get(key)); }

    @Override public Map<String, F> features() { return new HashMap<>(features); }
    @Override public Optional<F> feature(String key) { return Optional.ofNullable(features.get(key)); }

    @Override public StatIntermediate stats() { return stats; }
}

File: core/src/main/kotlin/com/gitlab/aecsocket/sokol/core/impl/AbstractComponent.kt
package com.gitlab.aecsocket.sokol.core.impl

import com.gitlab.aecsocket.glossa.core.I18N
import com.gitlab.aecsocket.sokol.core.DataNode
import com.gitlab.aecsocket.sokol.core.Feature
import com.gitlab.aecsocket.sokol.core.NodeComponent
import com.gitlab.aecsocket.sokol.core.Slot
import com.gitlab.aecsocket.sokol.core.rule.Rule
import com.gitlab.aecsocket.sokol.core.stat.ApplicableStats
import com.gitlab.aecsocket.sokol.core.stat.StatMap
import net.kyori.adventure.text.Component

abstract class AbstractComponent<
    C : AbstractComponent<C, F, S>,
    F : Feature.Profile<*>,
    S : Slot
>(
    override val id: String,
    override val tags: Set<String>,
    override val features: Map<String, F>,
    override val slots: Map<String, S>,
    override val stats: List<ApplicableStats>,
) : NodeComponent.Scoped<C, F, S> {
    override fun localize(i18n: I18N<Component>) =
        i18n.safe("component.$id")
}

open class SimpleSlot(
    override val key: String,
    override val tags: Set<String>,
    override val required: Boolean,
    val rule: Rule,
) : Slot {
    override fun localize(i18n: I18N<Component>) =
        i18n.safe("slot.$key")

    override fun compatible(node: DataNode) = rule.applies(node)
}
